src/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -612,23 +613,45 @@ public override bool Equals(object comparand)
 
         public override int GetHashCode()
         {
-            // For IPv6 addresses, we cannot simply return the integer
-            // representation as the hashcode. Instead, we calculate
-            // the hashcode from the string representation of the address.
+            if (_hashCode != 0)
+            {
+                return _hashCode;
+            }
+
+            // For IPv6 addresses, we calculate the hashcode by using Marvin
+            // on a stack-allocated array containing the Address bytes and ScopeId.
+            int hashCode;
             if (IsIPv6)
             {
-                if (_hashCode == 0)
+                const int addressAndScopeIdLength = IPAddressParserStatics.IPv6AddressBytes + sizeof(uint);
+                Span<byte> addressAndScopeIdSpan;
+                unsafe
                 {
-                    _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
+                    var addressSpan = new ReadOnlySpan<byte>(Unsafe.AsPointer(ref _numbers[0]), IPAddressParserStatics.IPv6AddressBytes);
+                    var scopeIdSpan = new ReadOnlySpan<byte>(Unsafe.AsPointer(ref _addressOrScopeId), sizeof(uint));
```

Comment:
@jkotas, is this safe?  I'd have thought the GC wouldn't scan the ptr resulting from Unsafe.AsPointer, but does it?

Reply:
This is not safe. This code has GC hole.

Reply:
I've used `new ReadOnlySpan<ushort>(_numbers).AsBytes()` instead. I assume that's safe as it doesn't go through a pointer, is it?

Reply:
Yes, that's fine.